substrate/client/network/sync/src/warp.rs

```diff
@@ -232,154 +303,242 @@ where
-					direction: Direction::Ascending,
-					max: Some(1),
-				};
-				Some((*header.number(), request))
-			},
+		let Phase::TargetBlock(header) = &mut self.phase else {
+			debug!(target: "sync", "Unexpected target block response from {peer_id}");
+			return Err(BadPeer(peer_id, rep::UNEXPECTED_RESPONSE))
+		};
+
+		if blocks.is_empty() {
+			debug!(
+				target: LOG_TARGET,
+				"Downloading target block failed: empty block response from {peer_id}",
+			);
+			return Err(BadPeer(peer_id, rep::NO_BLOCK))
 		}
-	}
 
-	/// Return target block hash if it is known.
-	pub fn target_block_hash(&self) -> Option<B::Hash> {
-		match &self.phase {
-			Phase::WarpProof { .. } | Phase::TargetBlock(_) | Phase::PendingTargetBlock { .. } =>
-				None,
-			Phase::State(s) => Some(s.target()),
+		if blocks.len() > 1 {
+			debug!(
+				target: LOG_TARGET,
+				"Too many blocks ({}) in warp target block response from {peer_id}",
+				blocks.len(),
+			);
+			return Err(BadPeer(peer_id, rep::NOT_REQUESTED))
+		}
+
+		validate_blocks::<B>(&blocks, &peer_id, Some(request))?;
+
+		let block = blocks.pop().expect("`blocks` len checked above; qed");
+
+		let Some(block_header) = &block.header else {
+			debug!(
+				target: "sync",
+				"Downloading target block failed: missing header in response from {peer_id}.",
+			);
+			return Err(BadPeer(peer_id, rep::VERIFICATION_FAIL))
+		};
+
+		if block_header != header {
+			debug!(
+				target: "sync",
+				"Downloading target block failed: different header in response from {peer_id}.",
+			);
+			return Err(BadPeer(peer_id, rep::VERIFICATION_FAIL))
 		}
+
+		if block.body.is_none() {
+			debug!(
+				target: "sync",
+				"Downloading target block failed: missing body in response from {peer_id}.",
+			);
+			return Err(BadPeer(peer_id, rep::VERIFICATION_FAIL))
+		}
+
+		self.result = Some(WarpSyncResult {
+			target_header: header.clone(),
+			target_body: block.body,
+			target_justifications: block.justifications,
+		});
+		self.phase = Phase::Complete;
+		self.actions.push(WarpSyncAction::Finished);
+		Ok(())
 	}
 
-	/// Return target block number if it is known.
-	pub fn target_block_number(&self) -> Option<NumberFor<B>> {
-		match &self.phase {
-			Phase::WarpProof { .. } | Phase::PendingTargetBlock { .. } => None,
-			Phase::TargetBlock(header) => Some(*header.number()),
-			Phase::State(s) => Some(s.target_block_num()),
+	/// Produce warp proof request.
+	fn warp_proof_request(&mut self) -> Option<(PeerId, WarpProofRequest<B>)> {
+		let Phase::WarpProof { last_hash, .. } = &self.phase else { return None };
+
+		// Copy `last_hash` early to cut the borrowing tie.
+		let begin = *last_hash;
+
+		if self
+			.peers
+			.iter()
+			.any(|(_, peer)| matches!(peer.state, PeerState::DownloadingProofs))
+		{
+			// Only one warp proof request at a time is possible.
+			return None
 		}
+
+		let Some((peer_id, peer)) = select_synced_available_peer(&mut self.peers, None) else {
+			return None
+		};
```

Comment:
```suggestion
		let (peer_id, peer) = select_synced_available_peer(&mut self.peers, None)?;
```

Reply:
I think the borrowing issue can be worked around the similar way as in state sync